### PR TITLE
Improve Adder API and perfomance

### PR DESCRIPTION
**Pull request by esanzgar**
Code reorganisation:
1. `Adder` has a simplified API: added a public method `display` that internally calls private methods `_target` and `_showAt`
2. Performance improvement in `_target` method by removing multiple calls to `this._width()` and `this._height()`
3. More strict type for `arrowDirection`

### CODE REVIEW

File: src/annotator/adder.js [hunk 1]
```diff
@@ -134,6 +142,24 @@ export class Adder {
     this._render();
   }
 
+  /**
+   * Display the adder in the best position in order to target the
+   * selected text in `targetRect`.
+   *
+   * @param {DOMRect} targetRect - The rect of text to target, in viewport
+   *        coordinates.
+   * @param {boolean} isSelectionBackwards - True if the selection was made
```
Comment (lyzadanger) [inline src/annotator/adder.js:151]: Just a random note: this assumes all displays are rendered left-to-right

Comment (esanzgar) [reply src/annotator/adder.js:151]: Could you let me know what make you think this assumes left-to-right?

Comment (lyzadanger) [reply src/annotator/adder.js:151]: If backwards means selecting right-to-left...that implies that the display is assumed to be left-to-right...

Maybe the word "backwards" is misleading here :)

File: src/annotator/adder.js [hunk 2]
```diff
@@ -4,24 +4,32 @@ import AdderToolbar from './components/adder-toolbar';
...
 
 /**
+ *  @typedef {2} ArrowPointingUp
  * Show the adder above the selection with an arrow pointing up at the
  * selected text.
  */
 export const ARROW_POINTING_UP = 2;
 
+/**
+ *  @typedef {ArrowPointingDown|ArrowPointingUp} ArrowDirection
```
Comment (lyzadanger) [inline src/annotator/adder.js:21]: This is helpful!

**Review by lyzadanger — CHANGES_REQUESTED** [review]
I like what's going on here!

As these changes are focused on the improvement of API and clarity, in large part, I've left some comments on how readability and organization could be tweaked yet further.

I think the one substantive change I'd request is the test you added to the `adder-test` module; see my inline comment for details.

File: src/annotator/adder.js [hunk 3]
```diff
@@ -134,22 +142,40 @@ export class Adder {
...
+   * Display the adder in the best position in order to target the
+   * selected text in `targetRect`.
+   *
+   * @param {DOMRect} targetRect - The rect of text to target, in viewport
+   *        coordinates.
+   * @param {boolean} isRTLselection - True if the selection was made
+   *        rigth-to-left, such that the focus point is mosty likely at the
+   *        top-left edge of `targetRect`.
+   */
+  display(targetRect, isRTLselection) {
```
[the remark this answers is not in the document]
Comment (esanzgar) [reply src/annotator/adder.js:155]: I agree

File: src/annotator/adder.js [hunk 4]
```diff
@@ -134,22 +142,40 @@ export class Adder {
...
+    const { left, top, arrowDirection } = this._target(
+      targetRect,
+      isRTLselection
+    );
+    this._showAt(left, top, arrowDirection);
```
[the remark this answers is not in the document]
Comment (esanzgar) [reply src/annotator/adder.js:160]: 👍🏼

File: src/annotator/test/adder-test.js [hunk 5]
```diff
@@ -256,12 +256,27 @@ describe('Adder', () => {
...
+
+  describe('#display', () => {
+    beforeEach(() => {
+      adderCtrl._target = sinon
+        .stub()
+        .returns({ left: 10, top: 10, arrowDirection: 2 });
+      adderCtrl._showAt = sinon.stub();
+    });
+
+    it('calls `_target` and `_showAt` methods', () => {
```
[the remark this answers is not in the document]
Comment (esanzgar) [reply src/annotator/test/adder-test.js:276]: 👍🏼

**Review by lyzadanger — APPROVED** [review]
Good work. I left one comment about a test if you could look at that; otherwise this is good to go.

File: src/annotator/test/adder-test.js [hunk 6]
```diff
@@ -256,12 +257,27 @@ describe('Adder', () => {
...
       });
     });
   });
+
+  describe('#show', () => {
+    it('shows the container in the correct location', () => {
+      adderCtrl.show(rect(100, 200, 100, 20), false);
+      const adder = document.elementFromPoint(150, 250);
+      assert.strictEqual(adder.label, 'adder-container');
+      assert.isTrue(+adder.style.zIndex > 0);
```
[the remark this answers is not in the document]
Comment (esanzgar) [reply src/annotator/test/adder-test.js:368]: It is intentional, it converts a numerical string "1" to a number. It is similar to `Number.parseInt()`

I think it is part of the `show` method:
* the element is positioned in the correct place
* the element shows on top of others (z-index)

**Pull request merged**